blog/2020-09-angular-schematics-common-helpers.md

```diff
@@ -682,11 +685,179 @@ export function playground(_options: any): Rule {
+  - Footer-Component (optional)
+- `ng new`: Create a new project with your company defaults
+  - Create the basic application generated from the Angular CLI (**`externalSchematic`**)
+  - Run the `ng add` schematic
+
+Alright, we already know how we can achieve the most of the things.
+But one thing we haven't learned yet: How to run other (external) Schematics?
+Therefore, we can use the `externalSchematic` function.
+
+But first things first, let's check if our collection file is ready to start:
+
+```json
+{
+  "$schema": "../node_modules/@angular-devkit/schematics/collection-schema.json",
+  "schematics": {
+    "ng-add": {
+      "description": "Call other Schematics from the same or other packages.",
+      "factory": "./ng-add/index#playground"
+    },
+    "ng-new": {
+      "description": "Execute `ng new` with predefined options and run other stuff",
+      "factory": "./ng-new/index#playground"
+    }
+  }
+}
+```
+
+> Using the special Schematic namen `ng-add` and `ng new` let's you later use the schematic with executing `ng add`/`ng new` (instead of other Schematics called with `ng generate`).
+> There is also the special Schematic name `ng-update`, that will be called in the end with the `ng update` Angular CLI command.
+
+Okay, now we can start implementing our Schematic.
+To execute an external Schematic, it must of course be available somewhere.
+But as we want to create a complete new project with your `ng new` Schematic, we don't have any `node_modules` installed in the target directory where we want to initialize the Angular workspace.
+We can use the [`spawn` method from `child_process`](https://nodejs.org/api/child_process.html#child_process_child_process_spawn_command_args_options) (available globally for node.js).
+This creates a new process that executes a command (in our case: `npm install @schematics/angular`).
+So make things sync we wrap the method call into a Promise and `await`ing for the Promise to be resolved.
+Now we listen to the `close` event from `spawn` and checking there was no error during install (code equals `0`).
+If everything worked fine, we will resolve the Promise.
+Otherwise, we can throw an error.
+The last step is to `chain` the `Rule`s we have.
+First we can now use the `externalSchematic` function to run the `ng new` Schematic from Angular itself and set up the basic app.
+We will hand over here some default options we always want to be set (such a using `SCSS`, support legacy browsers, strict mode, etc.).
+The second call is our `ng add` Schematic that adds our company specific components, UI libs and so on to the project.
+
+> We've already learned running a local Schematic by using the `RunSchematicTask` class that we add to our `context` object.
+> In this example we are using the `schematic` function to achieve the same goal.
+> Why are there two ways? To be honest: I actually don't know.
+> Checking the source code inside the Angular CLI, I found both implementations.
+
```

Comment:
```suggestion
After we defined the schema, we can now start to implement our Schematics.
To execute an external Schematic, it must be available in the scope of the project..
However, since we want to create a completely new project with the `ng new` Schematic, we don't have any `node_modules` installed in the target directory where we want to initialize the Angular workspace.
To run an external command we can use the [`spawn` method from `child_process`](https://nodejs.org/api/child_process.html#child_process_child_process_spawn_command_args_options) (available globally for Node.js).
This creates a new process that executes a command (in our case: `npm install @schematics/angular`).
To make things look synchronous we wrap the method call into a Promise and `await` for the Promise to be resolved.
Now we listen to the `close` event from `spawn` and check that there was no error during install (code equals `0`).
If everything worked fine, we will resolve the Promise, otherwise we can throw an error.
The last step is to `chain` all of our `Rule`s:
We first use the `externalSchematic()` function to run the `ng new` Schematic from Angular itself and set up the basic app.
We will hand over some default options here such a using `SCSS`, support legacy browsers, strict mode, etc.
The second call is our `ng add` Schematic that adds our company specific components, UI libs and so on to the project.

> We've already learned how to run a local Schematic by using the `RunSchematicTask` class that we need to add to our `context` object.
> In this example we are using the `schematic()` function to achieve the same goal.
> Why are there two ways? To be honest: I actually don't know.
> I found both implementations in the source code of Angular CLI.

```